sonarqube/datadog_checks/sonarqube/check.py

```diff
@@ -0,0 +1,194 @@
+        else:
+            self.service_check(self.SERVICE_CHECK_CONNECT, self.OK, tags=self._tags)
+
+    def collect_metrics(self):
+        available_metrics = self.discover_available_metrics()
+
+        for component, (tag_name, should_collect_metric) in self._components.items():
+            keys_to_query = []
+
+            for key, metric in available_metrics.items():
+                if should_collect_metric(metric):
+                    keys_to_query.append(key)
+
+            if not keys_to_query:
+                self.log.warning('Pattern for component `%s` does not match any available metrics', component)
+
+            response = self.http.get(
+                '{}/api/measures/component'.format(self._web_endpoint),
+                params={'component': component, 'metricKeys': ','.join(keys_to_query)},
+            )
+            response.raise_for_status()
+            metric_data = response.json()
+
+            for measure in metric_data['component']['measures']:
+                tags = ['{}:{}'.format(tag_name, component)]
+                tags.extend(self._tags)
+
+                self.gauge(available_metrics[measure['metric']], measure['value'], tags=tags)
+
+    def discover_available_metrics(self):
+        metadata_collected = False
+        available_metrics = {}
+
+        page = 1
+        seen = 0
+        total = -1
+
+        while seen != total:
+            response = self.http.get('{}/api/metrics/search'.format(self._web_endpoint), params={'p': page})
+            response.raise_for_status()
+
+            if not metadata_collected:
+                metadata_collected = True
+                self.collect_version(response)
+
+            search_results = response.json()
+            total = search_results['total']
+
+            for metric in search_results['metrics']:
+                seen += 1
+
+                if not self.is_valid_metric(metric):
+                    continue
+
+                domain = metric['domain']
+                key = metric['key']
+
+                category = CATEGORIES.get(domain)
+                if category is None:
+                    self.log.warning('Unknown metric category: %s', domain)
+                    continue
+
+                available_metrics[key] = '{}.{}'.format(category, key)
+
+            page += 1
+
+        return available_metrics
+
+    @AgentCheck.metadata_entrypoint
+    def collect_version(self, response):
+        version = response.headers.get('Sonar-Version', '')
+        if not version:
+            self.log.warning('The SonarQube version was not found in response headers')
+            return
+
+        # The version comes in like `8.5.0.37579` though sometimes there is no build part
+        version_parts = {name: part for name, part in zip(('major', 'minor', 'patch', 'build'), version.split('.'))}
+
+        self.set_metadata('version', version, scheme='parts', final_scheme='semver', part_map=version_parts)
+
+    def parse_config(self):
```

Comment:
Maybe add validation that JMX config options are not set for this instance?

Reply:
That would be a lot of options, are we sure?

Reply:
Maybe we can get check to ensure `host` and `port` are not set since those are required for JMX?